torch/csrc/jit/frontend/ir_emitter.cpp

```diff
@@ -1751,16 +1773,30 @@ struct to_ir {
     };
     GatheredTypes gathered(typeParser_);
     gathered.gather(classinfo);
-    auto val = emitExpr(obj);
+    Value* val = emitExpr(obj);
     RefinementSet refinement;
     if (gathered.types.size() == 1 &&
         gathered.types.at(0)->isSubtypeOf(val->type()) &&
         obj.kind() == TK_VAR) {
       std::string ident = Var(obj).name().name();
-      Refinement isinstance(std::move(ident), gathered.types.at(0));
-      refinement = RefinementSet({isinstance}, {});
+      Refinement isinstance(ident, gathered.types.at(0));
+
+      if (val->type()->kind() == UnionType::Kind) {
+        auto union_type = val->type()->expect<UnionType>();
+        std::vector<TypePtr> not_isinstance_types =
+            union_type->set_difference_of({gathered.types.at(0)});
```

Comment:
`set_difference_of` implies we are just subtracting types. I think in the use cases of the API in this file, what we want is to either take all types that are subtypes of a particular type, or take all types that are not subtypes of a particular type.